oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/MongoRegexPathFilterFactory.java

```diff
@@ -0,0 +1,149 @@
+        List<String> finalIncludedPathsRoots = new ArrayList<>();
+        LOG.debug("Candidate included paths: {}, candidate excluded paths: {}", sortedCandidateIncludedPaths, candidateExcludedPaths);
+        // The quadratic algorithm below could easily be replaced by linear time algorithm, but it's not worth the added
+        // complexity because the list of paths should never be more than a few tens
+        // Idea for linear time algorithm: after adding a path to the includedPathsRoots, advance in the list of candidates
+        // skipping all paths that are children of the path just added. Since the list is sorted alphabetically, the children
+        // of a path will be next to it in the list
+        for (String candidateIncludedPath : sortedCandidateIncludedPaths) {
+            if (finalIncludedPathsRoots.stream().noneMatch(ancestor -> PathUtils.isAncestor(ancestor, candidateIncludedPath))) {
+                finalIncludedPathsRoots.add(candidateIncludedPath);
+            }
+        }
+
+        List<String> finalExcludedPaths = new ArrayList<>();
+        // Keep an excluded path only if it is not needed by any filter.
+        for (String candidateExcludedPath : candidateExcludedPaths) {
+            boolean notNeeded = pathFilters.stream().allMatch(pathFilter -> pathFilter.filter(candidateExcludedPath) == PathFilter.Result.EXCLUDE);
+            if (notNeeded) {
+                finalExcludedPaths.add(candidateExcludedPath);
+            }
+        }
+
+        // Do not filter if there are too many includedPaths. The official Mongo documentation recommends
+        // not using more than tens of values in the $in filter to avoid performance degradation.
+        // https://www.mongodb.com/docs/manual/reference/operator/query/in/
+        if (finalIncludedPathsRoots.size() > maxNumberOfPaths) {
+            LOG.info("Number of includedPaths ({}) exceed the maximum allowed ({}), downloading everything",
+                    finalIncludedPathsRoots.size(), maxNumberOfPaths);
+            return MongoFilterPaths.DOWNLOAD_ALL;
+        }
+
+        // Also check the number of excluded paths and disable filtering by them if there are too many.
+        // We do not need to revert to download the full repository in the case where there is a reasonable number of
+        // included paths but too many excluded paths. The excluded paths are always children of the included paths, so
+        // if we do not filter on excludedPaths, we simply download everything inside the included paths. This might be
+        // much less than the full repository.
+        if (finalExcludedPaths.size() > maxNumberOfPaths) {
+            LOG.info("Number of excludedPaths ({}) exceed the maximum allowed ({}), disabling filtering of excludedPaths",
+                    finalExcludedPaths.size(), maxNumberOfPaths);
+            finalExcludedPaths = List.of();
+        }
+
+        if (finalExcludedPaths.isEmpty()) {
```

Comment:
I know this PR is not supposed to address this, but we do have the recommendation to not have either only `excludedPaths` or `includedPaths`. Not sure if it's helpful, but would logging a warning make sense here?

Reply:
We have documentation for having includedPaths (and queryPaths... but that's when running queries). We don't need documentation for excludedPaths in my view.